### PR TITLE
Introduce BuildFHSUserEnv with Bubblewrap as an alternative to chrootenv

### LINKED ISSUE #92798: Setting CAP_SYS_NICE=eip on vrcompositor-launcher crashes SteamVR
**Describe the bug**

SteamVR crashes when you set `CAP_SYS_NICE=eip` on vrcompositor-launcher.

**To Reproduce**
1. `sudo setcap CAP_SYS_NICE=eip ~/.local/share/Steam/steamapps/common/SteamVR/bin/linux64/vrcompositor-launcher`
2. `steam-run strace ~/.local/share/Steam/steamapps/common/SteamVR/bin/linux64/vrcompositor-launcher`

**Expected behavior**

On an FHS distro it looks for libs in /lib32 /lib64 and the like. It should either do that or look under /steamrt/.

**Screenshots**

**Additional context**

See #71554

I booted up a Pop!_OS live env, installed SteamVR, set the CAP and `strace`d vrcompositor-launcher.

It first started looking for libcap.so.2 in the SteamVR bin and then regular FHS paths like /usr/lib etc. where it found libcap under `/lib/x86_64-linux-gnu/libcap.so.2`.  
On NixOS it only tries to look in a few FHS paths under `${pkgs.glibc}` which is obviously nonsensical and fails.

Seems like this happends due to Nix-specific patches to dynamic linking because we're also doing something that makes vrcompositor-launcher try to preload from `/etc/ld-nix.so.preload` which, countrary to `LD_PRELOAD`, also actually works. I made use of this and modified `buildFHSUserEnv` to create a file in that location inside the chroot to preload libcap.so.2 and libstdc++.so.6 from the steamrt.  
With these patches, SteamVR can be launched with `CAP_SYS_NICE=eip` set on vrcompositor-launcher.

https://github.com/Atemu/nixpkgs/tree/steamvr-lib-loading-fixes

This is obviously not a solution, only for troubleshooting. Could someone more knowledgeable about dynamic linking in Nix to have a look at this now that we know what our issue is?

While doing this I have also discovered that the chronic stuttering issue (1-3 purple spikes ever second or so) is greatly reduced or even completely eliminated (depending on the mood of SteamVR) when a preloaded `${pkgs.steamPackages.steam-runtime}/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0` replaces the libSDL from the SteamVR bin it'd normally try to load. This can also be achieved by replacing the lib with a symlink to the one in the steamrt.  
I'd love to hear more reports on this so that we can take this information to Valve as I had the same issue on Pop!_OS.

**Notify maintainers**

@jagajaga @abbradar

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.6.19, NixOS, 20.09pre233323.dc80d7bc4a2 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.09pre233323.dc80d7bc4a2, nixpkgs-mozilla, nixos-stable-20.03.2310.fb6c3a6831c, nixos-unstable-20.09pre231796.22a81aa5fc1"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Atemu**
###### Motivation for this change

https://github.com/NixOS/nixpkgs/pull/55973#issuecomment-667068999

Fixes #92798

/cc @Mic92 @illegalprime 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:156]: If we are going for a new name, maybe we can find something that sounds less obscure.
I feel like adding a `Bw` suffix makes it a bit worse. On the other hand we could also just introduce this a new version of buildFHSUserEnv i.e. buildFHSUserEnv_2. At some point we should drop the old version.

Comment (Atemu) [reply pkgs/top-level/all-packages.nix:156]: As I said in the other PR, the name is supposed to be temporary. Once everything is ported, we would drop the suffix entirely.

Comment (Mic92) [reply pkgs/top-level/all-packages.nix:156]: Ok. I would still prefer a version number in this case to clearly communicate that.

Comment (Atemu) [reply pkgs/top-level/all-packages.nix]: I've made it clearer what is what now.

I also changed chrootenv -> fhsenv where applicable.

Comment (Mic92) [reply pkgs/top-level/all-packages.nix]: lgtm.

**Comment by Atemu** [comment]
I've come up with a cleaner solution to selectively use bubblewrap, will rebase to fix merge conflicts now

File: pkgs/build-support/build-fhs-userenv-bw/env.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Question: what is actually the purpose of shadowing etc?

Comment (Mic92) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Does this not make files such as `/etc/resolv.conf` inaccessible?

File: pkgs/top-level/all-packages.nix
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:156]:
```suggestion
  # TODO: migrate eventually everything to buildFHSUserEnvBubblewrap
  buildFHSUserEnv = buildFHSUserEnvChrootenv;
```

File: pkgs/build-support/build-fhs-userenv-bw/env.nix
Comment (Atemu) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Not sure (didn't write it) but that seems like a pretty straightforward way of bringing in a separate profile file and the mtab backcompat.  

I think that'd be a discussion for a separate issue/PR though.

Comment (Atemu) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Resolv.conf (or any other file in /etc) stays accessible btw, I believe this works like an overlay.

Comment (Mic92) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Ok. I was asking because @raboof was trying to remove it https://github.com/NixOS/nixpkgs/pull/80457 and I thought because we are now testing individual applications we might want to get rid of old stuff if it is not needed.

Comment (Mic92) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: The solution was to move it here: https://github.com/NixOS/nixpkgs/pull/80457/files#diff-0e0af17bff7614ce6822d64c6863bb1fR22

Comment (Atemu) [reply pkgs/build-support/build-fhs-userenv-bw/env.nix:72]: Sounds nice but I'd appteciate if we left that for another PR. This PR's one and only purpose is to reduce the scope of the change so that we can finally get it merged.

@raboof was actually waiting for the original PR to be merged and closed their PR for the mean time.

File: pkgs/top-level/all-packages.nix
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:158]:
```suggestion
  buildFHSUserEnvChroot = callPackage ../build-support/build-fhs-userenv { };
```

File: pkgs/top-level/all-packages.nix
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:157]:
```suggestion
  buildFHSUserEnv = buildFHSUserEnvChroot;
```

File: pkgs/build-support/build-fhs-userenv-bw/env.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/env.nix:102]:
```suggestion
  setupLibDirsTarget = ''
```

File: pkgs/build-support/build-fhs-userenv-bw/env.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/env.nix:109]:
```suggestion
  setupLibDirsMulti = ''
```

File: pkgs/build-support/build-fhs-userenv-bw/env.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/env.nix:128]:
```suggestion
  setupLibDirs = if isTargetBuild then setupLibDirsTarget
                                  else setupLibDirsMulti;
```

File: pkgs/build-support/build-fhs-userenv-bw/default.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/default.nix:61]:
```suggestion
  bwrapCmd = { init_args ? "" }: ''
```

File: pkgs/build-support/build-fhs-userenv-bw/default.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/default.nix:104]:
```suggestion
  bin = writeShellScriptBin name (bwrapCmd { initArgs = ''"$@"''; });
```

File: pkgs/build-support/build-fhs-userenv-bw/default.nix
Comment (Mic92) [inline pkgs/build-support/build-fhs-userenv-bw/default.nix:111]:
```suggestion
      shellHook = bwrapCmd {};
```

File: pkgs/top-level/all-packages.nix
Comment (Mic92) [inline pkgs/top-level/all-packages.nix:159]:
```suggestion
  buildFHSUserEnvBubblewrap = callPackage ../build-support/build-fhs-userenv-bubblewrap { };
```

**Comment by Atemu** [comment]
Sorry but this is just bikeshedding at this point. The FHSEnv infrastructure is in need of a refactoring but that's *WAY* out of scope for this PR.

Feel free to commit that stuff yourself (the branch is open) but honestly, it'd be better put in a separate PR. *After* we have deprecated chrootenv ideally.

**Review by Mic92 — CHANGES_REQUESTED** [review]
Sorry but consistent variable naming is not just bikeshedding. This is a builder and changing those variables later on can potentially break someones code.

**Comment by Atemu** [comment]
The variable names are consistent with the original build-fhs-chrootenv. Many of your suggestions are actually on variable names from the original. Take a look at the blame.

No, those names aren't what they should be but this PR is not a general refactoring of buildFHSUserEnv, variable names and code style. Intentionally so. Else it'd be going nowhere and end up with the same fate as the original PR. 
Perhaps it's an unavoidable fate though :/

**Comment by Atemu** [comment]
snake_case camelCase ;)

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 94442` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
<br>- steam-run-native
</details>
<details>
  <summary>10 packages built:</summary>
<br>- android-studio
<br>- androidStudioPackages.beta
<br>- androidStudioPackages.canary
<br>- androidStudioPackages.dev
<br>- kodiPlugins.steam-launcher
<br>- lutris
<br>- lutris-free
<br>- steam
<br>- steam-run
<br>- steamcmd
</details>

**Comment by Mic92** [comment]
Steam broke because of gstreamer, so it is unrelated.

**Comment by Mic92** [comment]
Steam broke because of gstreamer, so it is unrelated.

**Comment by Mic92** [comment]
@GrahamcOfBorg eval

**Comment by Atemu** [comment]
It's actually just the -native version, the regular (steamrt) steam is fine.

Was the nix-update hash change intended? That seems out of place.

**Comment by Mic92** [comment]
No sorry. I was testing nix-update with nix-update and it staged this file.

**Comment by Mic92** [comment]
Fixed via force push.

**Review by Mic92 — APPROVED** [review]
Good to merge after next eval.

**Comment by Atemu** [comment]
Thank you!

**Pull request merged**
